amethyst_network/src/simulation/transport/tcp.rs

```diff
@@ -242,32 +253,51 @@ impl<'s> System<'s> for TcpNetworkRecvSystem {
                 }
             };
 
-            loop {
-                match stream.read(&mut resource.recv_buffer) {
-                    Ok(recv_len) => {
-                        if recv_len > 0 {
-                            let event = NetworkSimulationEvent::Message(
-                                peer_addr,
-                                Bytes::copy_from_slice(&resource.recv_buffer[..recv_len]),
-                            );
-                            event_channel.single_write(event);
-                        } else {
+            // First get the pack size.
+            let mut len = vec![0; 4];
+            match stream.read(&mut len) {
+                Ok(tcp_pack_size) => {
+                    if tcp_pack_size == 0 {
+                        break;
+                    }
+                    let raw_size: [u8; 4] = [len[0], len[1], len[2], len[3]];
+                    pack_size = u32::from_le_bytes(raw_size.try_into().unwrap()) as usize;
+                }
+
+                Err(e) => {
+                    match e.kind() {
+                        io::ErrorKind::ConnectionReset => {
                             *active = false;
-                            break;
+                        }
+                        io::ErrorKind::WouldBlock => {}
+                        _ => {
+                            event_channel.single_write(NetworkSimulationEvent::RecvError(e));
                         }
                     }
-                    Err(e) => {
-                        match e.kind() {
-                            io::ErrorKind::ConnectionReset => {
-                                *active = false;
-                            }
-                            io::ErrorKind::WouldBlock => {}
-                            _ => {
-                                event_channel.single_write(NetworkSimulationEvent::RecvError(e));
-                            }
+                    break;
+                }
+            }
+
+            // Then grab this amount of data.
+            let mut data = vec![0; pack_size];
+            match stream.read(&mut data) {
```

Comment:
What if less than `pack_size` data is actually read? `tcp_pack_size` is not checked. You would have to keep some kind of recv buffer to reassemble incomplete packets between frames

Reply:
Did not think of this, would it be bad to block on the stream.read() until all the data is there?

If so I can implement a buffer.

Reply:
Yes blocking would be bad because any network lag would freeze the game loop so a buffer is needed.